File: asme/ops/serializers/teams.py
"""JSON shapes for teams (``/api/v1/teams``).

``team`` is the list/summary shape from the plan's serializer contract;
``team_detail`` adds the full membership list.
"""

from __future__ import annotations

from asme.ops.serializers import iso, project_ref, uid, user_ref


def _member_sort_key(member) -> tuple:
    user = member.user
    name = (user.display_name if user is not None else "").lower()
    return (0 if member.is_lead else 1, name, member.user_id)


def team(t) -> dict:
    members = list(t.members)
    leads = sorted((m for m in members if m.is_lead), key=_member_sort_key)
    return {
        "id": uid(t.id),
        "name": t.name,
        "description": t.description,
        "parent_team_id": uid(t.parent_team_id),
        "project": project_ref(t.project),
        "leads": [user_ref(m.user) for m in leads],
        "member_count": len(members),
        "is_active": bool(t.is_active),
        "created_at": iso(t.created_at),
        "updated_at": iso(t.updated_at),
    }


def team_member(m) -> dict:
    return {"user": user_ref(m.user), "is_lead": bool(m.is_lead), "joined_at": iso(m.joined_at)}


def team_detail(t) -> dict:
    data = team(t)
    data["members"] = [team_member(m) for m in sorted(t.members, key=_member_sort_key)]
    return data

File: apps/ops-web/src/api/contracts/teams.ts
import { z } from 'zod'
import { listOf, ProjectRef, UserRef } from './common'

export const Team = z.looseObject({
  id: z.string(),
  name: z.string(),
  description: z.string().nullable().optional(),
  parent_team_id: z.string().nullable().optional(),
  project: ProjectRef.nullable().optional(),
  leads: z.array(UserRef).default([]),
  member_count: z.number().optional(),
  is_active: z.boolean().optional(),
  created_at: z.string().optional(),
})
export type Team = z.infer<typeof Team>

export const TeamList = listOf(Team)

export interface TeamListParams {
  q?: string
  project?: string[]
  active?: 'true' | 'false'
  sort?: string
  limit?: number
  cursor?: string | null
}

/* ---- Teams / Users feature ------------------------------------------------- */

/** One entry of `team_detail.members` (asme/ops/serializers/teams.py::team_member). Leads come first. */
export const TeamMember = z.object({
  user: UserRef,
  is_lead: z.boolean(),
  joined_at: z.string().nullable().optional(),
})
export type TeamMember = z.infer<typeof TeamMember>

/** `GET|POST|PATCH /teams/:id` and `PUT /teams/:id/members` all answer with this shape under `payload.team`. */
export const TeamDetail = Team.extend({ members: z.array(TeamMember).default([]) })
export type TeamDetail = z.infer<typeof TeamDetail>

/** Create / edit form values; mirrors CREATE_SPEC / UPDATE_SPEC in asme/ops/services/teams.py. */
export const TeamInput = z.object({
  name: z.string().trim().min(1, 'Give the team a name.').max(160, 'Keep the name under 160 characters.'),
  description: z.string().trim().max(4000, 'Keep the description under 4,000 characters.').optional().or(z.literal('')),
  parent_team_id: z.string().nullable().optional(),
  project_id: z.string().nullable().optional(),
})
export type TeamInput = z.infer<typeof TeamInput>

/** One row of the `PUT /teams/:id/members` body. */
export interface TeamMemberInput {
  user_id: number
  is_lead: boolean
}

/** Body of `PUT /teams/:id/members`; replaces the whole membership. */
export interface TeamMembersInput {
  members: TeamMemberInput[]
}

/** Sorts accepted by `GET /teams` (services/teams.py::SORTS with the optional `-` prefix). */
export const TEAM_SORTS = ['name', '-name', 'created_at', '-created_at'] as const
export type TeamSort = (typeof TEAM_SORTS)[number]
